### PR TITLE
Port if, and, or, cond, and progn

### PR DESCRIPTION
**Pull request by shaleh**
Move original eval.rs to eval_macros.rs.

### CODE REVIEW

File: rust_src/src/eval.rs [hunk 1]
```diff
@@ -1,91 +1,100 @@
...
-            ::remacs_sys::message_with_string($str.as_ptr() as *const ::libc::c_char,
-                                              $obj.to_raw(),
-                                              $should_log);
+/// Eval args until one of them yields nil, then return nil.
+/// The remaining args are not evalled at all.
+/// If no arg yields nil, return the last arg's value.
+/// usage: (and CONDITIONS...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn and(args: LispObject) -> LispObject {
+    let mut val = Qnil;
```
Comment (Wilfred) [inline rust_src/src/eval.rs]: The default value should be Qt here. Would you mind adding a test too? `(and) => t`

File: rust_src/src/eval.rs [hunk 2]
```diff
@@ -1,91 +1,100 @@
...
+/// Returns the value of THEN or the value of the last of the ELSE's.
+/// THEN must be one expression, but ELSE... can be zero or more expressions.
+/// If COND yields nil, and there are no ELSE's, the value is nil.
+/// usage: (if COND THEN ELSE...)
+#[lisp_fn(name = "if", c_name = "if", min = "2", unevalled = "true")]
+pub fn lisp_if(args: LispObject) -> LispObject {
+    let cell = args.as_cons_or_error();
+    let cond = unsafe { eval_sub(cell.car().to_raw()) };
+
+    if cond == Qnil {
```
Comment (Wilfred) [inline rust_src/src/eval.rs]: This is putting the ELSE before the THEN. I think it's clearer with THEN first, which is what the C does.

File: rust_src/src/eval.rs (hunk 1 above)
Comment (shaleh) [reply rust_src/src/eval.rs]: Good catch on the `t` that is a little counter intuitive.

File: rust_src/src/eval.rs (hunk 2 above)
Comment (shaleh) [reply rust_src/src/eval.rs]: Will do

File: test/rust_src/src/eval-tests.el [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+(require 'ert)
+
+(ert-deftest eval-tests--or-base ()
+  "Check (or) base cases"
+  (should (eq (or) nil))
+  (should (eq (or t) t))
+  (should (eq (or nil) nil))
+  (should (eq (or nil t) t))
+  (should (eq (or t nil) t)))
```
Comment (Wilfred) [inline test/rust_src/src/eval-tests.el]: If we want to be thorough here, we could test non-booleans and evaluation order.

```
(or 1 2) => 1
(or t (error "oh no")) => t
```

That said, this is already much improved test coverage and I think the elisp tests would catch issues too.

File: test/rust_src/src/eval-tests.el [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+(ert-deftest eval-tests--cond-base ()
+  "Check (cond) base cases"
+  (should (eq (cond) nil))
+  (should (eq (cond (t)) t))
+  (should (eq (cond (nil)
+                    (t))
+              t))
+  (should (eq (cond (t)
+                    (nil))
+              t)))
```
Comment (Wilfred) [inline test/rust_src/src/eval-tests.el]: Clauses can take two forms with cond:

```
(cond (1)) => 1
(cond (1 2 3)) => 3
```

We could also test evaluation here too:

```
(cond (nil (error "oh no"))
  (t 1)) => 1
```

File: test/rust_src/src/eval-tests.el (hunk 3 above)
Comment (shaleh) [reply test/rust_src/src/eval-tests.el]: Good idea.

File: test/rust_src/src/eval-tests.el (hunk 4 above)
Comment (shaleh) [reply test/rust_src/src/eval-tests.el]: roger that

**Pull request merged**